### PR TITLE
OpenAL replacement

### PR DESCRIPTION
**Pull request by NiceGuyAndersson**
Can currently play several audio sources simultaneously in mono. Positional stereo to be added later.

### CODE REVIEW

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/e3657921c838d8c9fca4e2b5a66f67fdd3b5953f/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L141](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/e3657921c838d8c9fca4e2b5a66f67fdd3b5953f/src/Engine/Manager/SoundManager.cpp#L141): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/4ce9816eaa0ebd6b9a00987d1d6407deffc78302/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L141](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/4ce9816eaa0ebd6b9a00987d1d6407deffc78302/src/Engine/Manager/SoundManager.cpp#L141): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Audio/SoundBuffer.hpp [hunk 1]
```diff
@@ -45,14 +45,22 @@ namespace Audio {
              * @param soundFile The sound file containing the sound.
              */
             void Load(SoundFile* soundFile);
-            
+
+            /// Get size of buffer.
+            /**
+             * @return size_t The size of the buffer.
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SoundBuffer.hpp]: Type of the return parameter should be document, just its function. Doxygen can already determine the type of parameters.

File: src/Engine/Audio/SoundBuffer.hpp [hunk 2]
```diff
@@ -45,14 +45,22 @@ namespace Audio {
...
+             */
+            size_t GetSize();
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SoundBuffer.hpp]: size_t is in the std namespace

File: src/Engine/Audio/SoundFile.hpp [hunk 3]
```diff
@@ -16,25 +16,18 @@ namespace Audio {
...
              */
-            virtual ALsizei GetSize() const = 0;
-            
-            /// Get AL format.
-            /**
-             * 32-bit sound is not supported in OpenAL.
-             * @return One of AL_FORMAT_MONO8, AL_FORMAT_MONO16, AL_FORMAT_STEREO8 or AL_FORMAT_STEREO16.
-             */
-            virtual ALenum GetFormat() const = 0;
+            virtual size_t GetSize() const = 0;
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SoundFile.hpp]: size_t is in the std namespace

File: src/Engine/Audio/SoundFile.hpp [hunk 4]
```diff
@@ -16,25 +16,18 @@ namespace Audio {
...
             
             /// Get sample rate.
             /**
              * @return The sound file's sample rate (Hz).
              */
-            virtual ALsizei GetSampleRate() const = 0;
+            virtual size_t GetSampleRate() const = 0;
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SoundFile.hpp]: size_t is in the std namespace

File: src/Engine/Audio/SteamAudio.cpp [hunk 5]
```diff
@@ -1 +1,57 @@
...
+SteamAudio::SteamAudio(IPLContext * context, IPLhandle * environment)
+{
+}
+
+SteamAudio::~SteamAudio()
+{
+}
+
+void SteamAudio::Process(IPLAudioBuffer input, IPLVector3 * playerPos, IPLVector3 * playerDir, IPLVector3 * playerUp, IPLVector3 * sourcePos, float sourceRadius)
+{
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.cpp]: Brackets on same line

File: src/Engine/Audio/SteamAudio.cpp [hunk 6]
```diff
@@ -1 +1,57 @@
...
+
+    processedBuffers.push_back(input);
+
+}
+
+void SteamAudio::GetFinalMix(IPLAudioBuffer* finalBuf, size_t* numSamples)
+{
+    // No audio is playing, return emptiness, let SoundManager figure it out.
+    if (processedBuffers.size() == 0)
+    {
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.cpp]: Brackets on same line

File: src/Engine/Audio/SteamAudio.hpp [hunk 7]
```diff
@@ -27,12 +28,23 @@ class SteamAudio {
...
         */
-        IPLAudioBuffer Process(IPLAudioBuffer input, IPLVector3* playerPos, IPLVector3* playerDir, IPLVector3* playerUp, IPLVector3* sourcePos, float sourceRadius);
+        void Process(IPLAudioBuffer input, IPLVector3* playerPos, IPLVector3* playerDir, IPLVector3* playerUp, IPLVector3* sourcePos, float sourceRadius);
+
+        /// Mixes and returns the final buffer, ready to be played.
+        /**
+         * @param finalBuf Pointer to an empty buffer.
+         * @param numSamples The number of samples in the final buffer.
+         */
+        void GetFinalMix(IPLAudioBuffer* finalBuf, size_t* numSamples);
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.hpp]: size_t is in the std namespace

File: src/Engine/Audio/SteamAudio.hpp [hunk 8]
```diff
@@ -27,12 +28,23 @@ class SteamAudio {
         * @param sourceRadius The radius of the source, for calculating occlusion.
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.hpp]: Non-existent parameter

File: src/Engine/Audio/SteamAudio.hpp [hunk 9]
```diff
@@ -27,12 +28,23 @@ class SteamAudio {
...
         * @return The processed buffer.
         */
-        IPLAudioBuffer Process(IPLAudioBuffer input, IPLVector3* playerPos, IPLVector3* playerDir, IPLVector3* playerUp, IPLVector3* sourcePos, float sourceRadius);
+        void Process(IPLAudioBuffer input, IPLVector3* playerPos, IPLVector3* playerDir, IPLVector3* playerUp, IPLVector3* sourcePos, float sourceRadius);
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudio.hpp]: Undocumented parameters

File: src/Engine/Audio/SteamAudioDirectRenderer.cpp [hunk 10]
```diff
@@ -1 +1,13 @@
 #include "SteamAudioDirectRenderer.hpp"
+
+SteamAudioDirectRenderer::SteamAudioDirectRenderer()
+{
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudioDirectRenderer.cpp]: Brackets on same line

File: src/Engine/Audio/SteamAudioIndirectRenderer.cpp [hunk 11]
```diff
@@ -1 +1,13 @@
 #include "SteamAudioIndirectRenderer.hpp"
+
+SteamAudioIndirectRenderer::SteamAudioIndirectRenderer()
+{
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudioIndirectRenderer.cpp]: Brackets on same line

File: src/Engine/Audio/SteamAudioInterface.cpp [hunk 12]
```diff
@@ -1,2 +1,35 @@
...
+
+float * SteamAudioInterface::GetProcessed(size_t* numSamples) {
+    IPLAudioBuffer* finalBuf = new IPLAudioBuffer;
+    sAudio.GetFinalMix(finalBuf, numSamples);
+
+    return finalBuf->interleavedBuffer;
+}
+
+SteamAudioInterface::SteamAudioInterface()
+{
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudioInterface.cpp]: Brackets on same line

File: src/Engine/Audio/SteamAudioInterface.hpp [hunk 13]
```diff
@@ -102,7 +102,14 @@ class SteamAudioInterface {
...
          **/
-        void Process(IPLAudioBuffer input, IPLVector3* sourcePosition, float sourceRadius); // Needs to be called in a way so that there's always at least one processed audio frame ready to go.
+        void Process(float* input, size_t samples, IPLVector3* sourcePosition, float sourceRadius); // Needs to be called in a way so that there's always at least one processed audio frame ready to go.
+
+        /// Fetches the processed mix using all buffers sent through Process()
+        /**
+         * @param numSamples The number of samples in the final buffer.
+         * @return A pointer to the processed buffer.
+         **/
+        float* GetProcessed(size_t* numSamples);
```
Comment (Chainsawkitten) [inline src/Engine/Audio/SteamAudioInterface.hpp]: size_t is in the std namespace

File: src/Engine/Audio/VorbisFile.hpp [hunk 14]
```diff
@@ -22,29 +22,22 @@ namespace Audio {
...
              */
-            ALsizei GetSize() const final;
-            
-            /// Get AL format.
-            /**
-             * 32-bit sound is not supported in OpenAL.
-             * @return One of AL_FORMAT_MONO8, AL_FORMAT_MONO16, AL_FORMAT_STEREO8 or AL_FORMAT_STEREO16.
-             */
-            ALenum GetFormat() const final;
+            size_t GetSize() const final;
```
Comment (Chainsawkitten) [inline src/Engine/Audio/VorbisFile.hpp]: size_t is in the std namespace

File: src/Engine/Component/SoundSource.hpp [hunk 15]
```diff
@@ -46,11 +46,14 @@ namespace Component {
             
             /// Whether the sound should loop.
             bool loop = false;
+
+            /// Place in buffer
+            size_t place = 0;
```
Comment (Chainsawkitten) [inline src/Engine/Component/SoundSource.hpp]: size_t is in the std namespace

File: src/Engine/Manager/SoundManager.hpp [hunk 16]
```diff
@@ -20,10 +21,10 @@ class SoundManager {
         /**
          * @param message Message to print to standard error if an error was encountered.
          */
-        static void CheckError(const char* message);
+        static void CheckError(PaError err);
         
         /// Moves sound sources and plays sounds.
-        void Update();
+        void Update(float deltaTime);
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SoundManager.hpp]: Undocumented parameter

File: src/Engine/Manager/SoundManager.hpp [hunk 17]
```diff
@@ -20,10 +21,10 @@ class SoundManager {
         /**
          * @param message Message to print to standard error if an error was encountered.
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SoundManager.hpp]: Non-existent parameter

File: src/Engine/Manager/SoundManager.hpp [hunk 18]
```diff
@@ -20,10 +21,10 @@ class SoundManager {
...
          */
-        static void CheckError(const char* message);
+        static void CheckError(PaError err);
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SoundManager.hpp]: Undocumented parameter

File: src/Engine/Manager/SoundManager.hpp [hunk 19]
```diff
@@ -1,7 +1,8 @@
 #pragma once
 
-#include <AL/alc.h>
 #include "../Entity/ComponentContainer.hpp"
+#include "portaudio.h"
```
Comment (Chainsawkitten) [inline src/Engine/Manager/SoundManager.hpp]: Use <> for non-relative includes.

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/db449a9fa773d46ba1478a08ebc78a57a233c860/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L141](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/db449a9fa773d46ba1478a08ebc78a57a233c860/src/Engine/Manager/SoundManager.cpp#L141): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d2c75d9fb891bd231d8ebc61cfdcb9897f626af3/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L141](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d2c75d9fb891bd231d8ebc61cfdcb9897f626af3/src/Engine/Manager/SoundManager.cpp#L141): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/58da0f3ac11f2510b600ae3a6cdecad1dca6475c/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L141](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/58da0f3ac11f2510b600ae3a6cdecad1dca6475c/src/Engine/Manager/SoundManager.cpp#L141): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1d403fc7e827cf97802744717c97059876450ce4/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L141](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1d403fc7e827cf97802744717c97059876450ce4/src/Engine/Manager/SoundManager.cpp#L141): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Audio/SteamAudio.hpp (hunk 8 above)
Comment (NiceGuyAndersson) [reply src/Engine/Audio/SteamAudio.hpp]: I think you're mistaken?

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/675d098afbceba725eda53606df1774f58c761dd/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L141](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/675d098afbceba725eda53606df1774f58c761dd/src/Engine/Manager/SoundManager.cpp#L141): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [Hymn.cpp#L172](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/c20f00b704ac9dea560728312772684278bd320a/src/Engine/Hymn.cpp#L172): Variable 'model' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [SoundManager.cpp#L141](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/c20f00b704ac9dea560728312772684278bd320a/src/Engine/Manager/SoundManager.cpp#L141): The function 'GetListeners' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**